File: Nebokrai/server/src/billing/appbilling/servermessage.cpp
#include "servermessage.h"

#include "../../nets/netbilling/message.h"
#include "../../nets/netbilling/serverforgs.h"

#include <spdlog/spdlog.h>

#include <array>
#include <bit>
#include <string>
#include <vector>

namespace Billing
{
namespace
{
constexpr std::int32_t kGameServerConnected = 0x000E'F101;
constexpr std::int32_t kGameServerDisconnected = 0x0010'EF01;
constexpr std::size_t kAddressLimit = 0x20;

std::string BytesToString(std::span<const std::uint8_t> bytes)
{
    return {bytes.begin(), bytes.end()};
}

std::string FormatIPv4(std::uint32_t address)
{
    const auto bytes = std::bit_cast<std::array<std::uint8_t, 4>>(address);
    return std::to_string(bytes[0]) + "." + std::to_string(bytes[1]) + "." +
           std::to_string(bytes[2]) + "." + std::to_string(bytes[3]);
}
}

ServerMessageOutcome OnServerMessage(BillingNet::CMessage& message,
                                     BillingNet::CServerForGS* server)
{
    if (message.MessageType() == kGameServerConnected) {
        if (server != nullptr) {
            static_cast<void>(server->SetClientMapID(message.SocketID(), message.SocketID()));
        }
        spdlog::info("BillingServer: GameServer {} подключён с адреса {}",
                     message.SocketID(), FormatIPv4(message.IP()));
        return {.kind = ServerMessageOutcome::Kind::Connected};
    }

    if (message.MessageType() == kGameServerDisconnected) {
        const std::int32_t mapId = message.Base().GetLong();
        auto addressBytes = message.Base().GetStrBytes(kAddressLimit)
                                .value_or(std::vector<std::uint8_t>{});
        const std::int32_t port = message.Base().GetLong();
        const bool allowed = server != nullptr &&
            server->IsAllowedAddress(addressBytes, static_cast<std::uint16_t>(port));
        spdlog::info("BillingServer: {}GameServer {}<{}:{}> отключён",
                     allowed ? "" : "неразрешённый ",
                     mapId,
                     BytesToString(addressBytes),
                     port);
        return {.kind = ServerMessageOutcome::Kind::Disconnected, .allowed = allowed};
    }
    return {};
}
}

File: Nebokrai/server/rust/src/billingserver/appbilling/servermessage.rs
//! GameServer lifecycle-ветви `appbilling/servermessage.cpp`, подтверждённые
//! `billingserver.exe` и `billingserver.pdb`.
//!
//! Connect сначала назначает socket как map ID, затем публикует событие.
//! Disconnect всегда проверяет переданный адрес и `port as u16`, независимо от
//! admission-флага allow-list; signed payload сохраняется, хотя старый журнал
//! показывал те же биты как unsigned. Nullable server-owner выражен `Option`.

use std::net::Ipv4Addr;

use crate::nets::netbilling::message::CMessage;
use crate::nets::netbilling::serverforgs::CServerForGS;

const GAME_SERVER_CONNECTED_MESSAGE_TYPE: i32 = 0x000E_F101;
const GAME_SERVER_DISCONNECTED_MESSAGE_TYPE: i32 = 0x0010_EF01;
const GAME_SERVER_ADDRESS_LIMIT: usize = 0x20;

#[derive(Debug, Eq, PartialEq)]
pub(crate) enum ServerMessageOutcome {
    Connected {
        socket_id: i32,
        address: Ipv4Addr,
        map_assignment: Option<i32>,
    },
    Disconnected {
        map_id: i32,
        address: Vec<u8>,
        port: i32,
        allowed: bool,
    },
    Unsupported { message_type: i32 },
}

pub(crate) struct ServerMessageHandler<'a> {
    server: Option<&'a CServerForGS>,
}

impl<'a> ServerMessageHandler<'a> {
    pub(crate) const fn new(server: Option<&'a CServerForGS>) -> Self {
        Self { server }
    }

    pub(crate) fn on_server_message(&self, message: &mut CMessage) -> ServerMessageOutcome {
        match message.message_type() {
            GAME_SERVER_CONNECTED_MESSAGE_TYPE => self.on_connected(message),
            GAME_SERVER_DISCONNECTED_MESSAGE_TYPE => self.on_disconnected(message),
            message_type => ServerMessageOutcome::Unsupported { message_type },
        }
    }

    fn on_connected(&self, message: &CMessage) -> ServerMessageOutcome {
        let socket_id = message.socket_id();
        let map_assignment = self.server.map(|server| {
            server
                .command_handle()
                .set_client_map_id(socket_id, socket_id)
        });
        let address = Ipv4Addr::from(message.ip().to_le_bytes());

        ServerMessageOutcome::Connected {
            socket_id,
            address,
            map_assignment,
        }
    }

    fn on_disconnected(&self, message: &mut CMessage) -> ServerMessageOutcome {
        let map_id = message.base_mut().get_long().unwrap_or(0);
        let address = message
            .base_mut()
            .get_str_bytes(GAME_SERVER_ADDRESS_LIMIT)
            .expect("ненулевая GetStr-граница задана константой");
        let port = message.base_mut().get_long().unwrap_or(0);
        let allowed = self
            .server
            .is_some_and(|server| server.is_allowed_address(&address, port as u16));

        ServerMessageOutcome::Disconnected {
            map_id,
            address,
            port,
            allowed,
        }
    }
}
